src/main/java/com/github/nishgpt/chainexecutor/StageExecutionManager.java

```diff
@@ -43,28 +43,31 @@ public U execute(StageExecutorKey<T, K> stageExecutorKey, C chainIdentifier, U c
       final var executor = executorFactory.getExecutor(stageExecutorKey);
       final var currentStatus = executor.getStageStatus(context);
 
-      //Init this stage if not initiated
+      //Init this stage if not initiated and can't be skipped
       if (currentStatus.isNotInitiated()) {
         log.info("Initiating {} Stage for id - {}", stageExecutorKey.getStage(), context.getId());
-        context = (U) executor.init(context);
+        final var updatedContext = (U) executor.skipIfApplicable(context);
+        context = executor.getStageStatus(updatedContext).isSkipped() ? updatedContext : (U) executor.init(updatedContext);
       }
 
       //If the stage is already Failed, skip execution
       if (currentStatus.isFailed()) {
         return context;
       }
 
-      //Execute this stage
-      log.info("Executing {} Stage for id - {}", stageExecutorKey.getStage(), context.getId());
-      var executedContext = (U) executor.execute(context, request);
+      if (!currentStatus.isSkipped()) {
+        //Execute this stage
+        log.info("Executing {} Stage for id - {}", stageExecutorKey.getStage(), context.getId());
+        context = (U) executor.execute(context, request);
+      }
 
       //If execution is not completed, expect a call to resume flow
-      if (!executor.getStageStatus(executedContext).isCompleted()) {
-        return executedContext;
+      if (executor.getStageStatus(context).isExecutable()) {
```

Comment:
Lets use the earlier version of isNotCompleted, which is more intuitive that we will resume it later. Skipped anyways wont be a scenario here